content/blog/safe-rust-isnt-always-safe/index.md

```diff
@@ -0,0 +1,692 @@
+- [Don't Index Into Arrays Without Bounds Checking](#don-t-index-into-arrays-without-bounds-checking)
+- [Use `split_at_checked` Instead Of `split_at`](#use-split-at-checked-instead-of-split-at)
+- [Make Invalid States Unrepresentable](#make-invalid-states-unrepresentable)
+- [Ensure Thread-Safety](#ensure-thread-safety)
+- [Avoid Primitive Types For Business Logic](#avoid-primitive-types-for-business-logic)
+- [Handle Default Values Carefully](#handle-default-values-carefully)
+- [Implement `Debug` Safely](#implement-debug-safely)
+- [Implement Secure Serialization](#implement-secure-serialization)
+- [Protect Against Time-of-Check to Time-of-Use (TOCTOU)](#protect-against-time-of-check-to-time-of-use-toctou)
+- [Use Constant-Time Comparison for Sensitive Data](#use-constant-time-comparison-for-sensitive-data)
+- [Don't Accept Unbounded Input](#don-t-accept-unbounded-input)
+- [Path `join` Swallows Errors](#path-join-swallows-errors)
+- [Check For Unsafe Code In Your Dependencies With `cargo-geiger`](#check-for-unsafe-code-in-your-dependencies-with-cargo-geiger)
+- [Conclusion](#conclusion)
+
+</details>
+
+## Protect Against Integer Overflow
+
+Overflow errors can happen pretty easily:
+
+```rust
+// DON'T: Use unchecked arithmetic
+fn calculate_total(price: u32, quantity: u32) -> u32 {
+    price * quantity  // Could overflow!
+}
+```
+
+If `price` and `quantity` are large enough, the result will overflow.
+Rust will panic in debug mode, but in release mode, it will silently wrap around.
+
+To avoid this, use checked arithmetic operations:
+
+```rust
+// DO: Use checked arithmetic operations
+fn calculate_total(price: u32, quantity: u32) -> Result<u32, ArithmeticError> {
+    price.checked_mul(quantity)
+        .ok_or(ArithmeticError::Overflow)
+}
+```
+
+Note that static checks are not removed because they don't compromise on performance of generated code. 
+If the compiler is able to detect the problem at compile time, it will do so:
+
+```rust
+fn main() {
+    let x: u8 = 2;
+    let y: u8 = 128;
+    let z = x * y;  // Compile-time error!
+}
+```
+
+The error message will be:
+
+```rust
+error: this arithmetic operation will overflow
+ --> src/main.rs:4:13
+  |
+4 |     let z = x * y;  // Compile-time error!
+  |             ^^^^^ attempt to compute `2_u8 * 128_u8`, which would overflow
+  |
+  = note: `#[deny(arithmetic_overflow)]` on by default
+```
+
+For all other cases, use [`checked_add`](https://docs.rs/num/latest/num/trait.CheckedAdd.html), [`checked_sub`](https://docs.rs/num/latest/num/trait.CheckedSub.html), [`checked_mul`](https://docs.rs/num/latest/num/trait.CheckedMul.html), and [`checked_div`](https://docs.rs/num/latest/num/trait.CheckedDiv.html) to avoid overflow.
+
+{% info(title="Quick Tip: Enable Overflow Checks In Release Mode", icon="info") %}
+
+In general, Rust values performance at least as much as safety.
+Overflow checks are expensive, which is why Rust disables them in release mode.
```

Comment:
> Overflow checks are expensive

Might be worth backing this up with some evidence. Here's an older post from Dan Luu https://danluu.com/integer-overflow/ which suggests the checking isn't too bad but it can have an impact on other optimisations.

Reply:
Thanks! It's probably outside of the scope of this post, but I'll add the link for further reference and change the wording to "can be expensive."